src/votekit/outcome_model.py

```diff
@@ -0,0 +1,140 @@
+            l = self.eliminated.copy()
+            l.reverse()
+            return l
+    def get_rankings(self) -> list[str]:
+        return self.get_all_winners() + self.remaining +  self.get_all_eliminated()
+    
+    def get_profile(self) -> PreferenceProfile:
+        if self.profile != None:
+            return self.profile
+        elif self.previous==None:
+            raise ValueError(f"No profile found")
+        else:
+            return self.previous.get_profile()
+    
+    def get_round_outcome(self,roundNum: int) -> list[str]:
+        if self.curr_round == roundNum:
+            # current round's eliminated and/or elected
+            if self.elected != []:
+                return self.elected
+            else:
+                return self.eliminated
+        else:
+            self.previous.get_round_outcome(roundNum)
+    
+  ###############################################################################################
+    
+    """def add_winners_and_losers(self, winners: set[str], losers: set[str]) -> "Outcome":
+        # example method, feel free to delete if not useful
+        if not winners.issubset(self.remaining) or not losers.issubset(self.remaining):
+            missing = (winners.difference(set(self.remaining)) | (losersdifference(set(self.remaining)))
+            raise ValueError(f"Cannot promote winners, {missing} not in remaining")
+        return Outcome(
+            remaining=set(self.remaining).difference(winners | losers),
+            elected=list(set(self.elected) | winners)
+            eliminated=list(set(self.eliminated) | losers)
+        )
+    """
+    
+
+    
+    # THIS ASSUMES BALLOTS DON'T CHANGE BY ROUND!
+    def difference_remaining_candidates(self, prevOutcome1: PreferenceProfile, prevOutcome2: PreferenceProfile) -> float:
+        """
+        Returns the fractional difference in number of remaining candidates
+        """
+        
+        # check if from same contest
+        if set(prevOutcome1.get_profile().ballots) != set(prevOutcome2.get_profile().ballots):
+            raise ValueError("Cannot compare outcomes from different elections")
+        
+        remaining_diff = len((prevOutcome1.remaining).difference(prevOutcome2.remaining))
+        allcandidates = len(prevOutcome1.remaining) + len(prevOutcome1.get_all_winners()) + len(prevOutcome1.get_all_eliminated())
+        return remaining_diff / allcandidates
+    
+    def changed_rankings(self) -> dict: # {str candidate : (oldRank, newRank)}
+        """
+        Returns dict of (key) string candidates who changed ranking from previous round and (value) a tuple of (prevRank, newRank)
+        """
+        if self.previous == None:
+            raise ValueError("This is the first round, cannot compare previous ranking")
+        
+        else:
+            prev_ranking = self.previous.get_rankings()
+            curr_ranking = self.get_rankings()
+            if curr_ranking == prev_ranking:
+                return {}
```

Comment:
Let's try and avoid returning empty dictionaries, maybe this is a print(statement) that states the current and previous rankings are the same

Reply:
gotcha